### PR TITLE
Initial UnusedReferenceService API

### PR DESCRIPTION
**Pull request by JoeRobich**
This adds the types needed for the Project System to implement their share of the feature's logic.

- [x] Need to determine whether an IVT needs to be added and whether to add it for Microsoft.ProjectSystem.Managed or Microsoft.ProjectSystem.Managed.VS

CC: @ocallesp 

The goal of this api is to allow a UI dialog and a compilation end analyzer to be written. The UI dialog would be invoked from the Solution Explorer context menu for project and solution nodes and provide the user with the ability to remove unused references. The analyzer would during a build and emit diagnostics based on the presense of unused references.

The IUnusedReferenceService would be the Roslyn entry point to this analysis and provide the means to update these unused references.
The IReferenceUpdateService would be implemented by the Project System and MEF imported to allow us to query information from the build and to update references in project files.

### CODE REVIEW

File: src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace Microsoft.CodeAnalysis.UnusedReferences
+{
+    internal interface IUnusedReferencesService
+    {
+        /// <summary>
+        /// Determines unused references from this compilation.
+        /// </summary>
+        Task<ImmutableArray<UnusedReference>> GetUnusedReferencesAsync(
+            Project project,
+            Compilation compilation,
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:18]: Why does this need a compilation of it already takes a project?

File: src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+        /// <summary>
+        /// Updates references by removing, adding, or marking references as TreatAsUsed
+        /// for this project.
+        /// </summary>
+        /// <returns>Updated project</returns>
+        Task<Project> UpdateReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceUpdate> referenceUpdates,
+            CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:29]: Not clear to me yet what this is needed for

File: src/Features/Core/Portable/UnusedReferences/Reference.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        /// </summary>
+        public ReferenceType Type { get; }
+
+        /// <summary>
+        /// Uniquely identifies the reference.
+        /// </summary>
+        /// <remarks>
+        /// Should match the Include or Name attribute used in the project file.
+        /// </remarks>
+        public string ItemSpec { get; }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/Reference.cs:20]: Specification vs spec?

File: src/Features/Core/Portable/UnusedReferences/UpdateAction.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        /// <summary>
+        /// No action needs to be performed.
+        /// </summary>
+        None,
+
+        /// <summary>
+        /// Indicates the reference needs to be updated.
+        /// </summary>
+        Update,
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/UpdateAction.cs:17]: What's an example of this?

Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/UpdateAction.cs:17]: One of the goals is to have the ability to tag references as "TreatAsUsed" (or some similar attribute) to indicate that although the reference is unused at compile time, it should be considered used for the purposes of analyzing unused references. This update action would be sent to the project system after a user has changed an unused reference's treated as used state.

File: src/Features/Core/Portable/UnusedReferences/Reference.cs (hunk 3 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/Reference.cs:20]: I believe the MSBuild term for this concept is ItemSpec, but I have no problem with ItemSpecification.

File: src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs (hunk 2 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:29]: The Unused Reference UI would invoke this to apply the changes the user has choosen. Removing direct references, adding transitive references, or updating a TrestAsUsed reference attribute.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:29]: a few more questions:

1. "Removing direct references".  If they have a Project instance, why can't htey just do the removal themselve and give us back the new Project instance?
2. "or updating a TrestAsUsed reference attribute."  I feel like we want a bit on MetadataREference to say that it is TreatAsUsed.  With that bit all hte UI has to do is fork teh Project instance and replace the MetadataReferences with teh appropriate forked versions with that bit set.
3. "adding transitive references,".  Can you expand on this.  Why is that part of an 'unused reference ui'?

Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:29]: @CyrusNajmabadi Thanks for your questions. My goal for this service is that it provides the high-level Project and Package references that are unused and allows those high-level references to be updated.

1. So, the UI will invoke this method with the list of changes that the user requested. This method will then work with the project system to ensure that the Project and Package references are updated accordingly.
2. A single Project or Package references is likely composed of many MetadataReferences. I am also unsure how an updated metadata reference results in an updated project file. Unless you are saying each of these updates could be encoded as changes in a project file.
3. It is possible that a no types from a directly referenced Project or Package are used in the compilation, but types from a tansitive reference it brings along are used. We could provide a means to remove the unused direct reference and directly add the transitive reference in these cases.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:29]: > It is possible that a no types from a directly referenced Project or Package are used in the compilation, but types from a tansitive reference it brings along are used. We could provide a means to remove the unused direct reference and directly add the transitive reference in these cases.

i don't think that's possible... can you give an example.  i believe the compielr needs all references to be direct to understand anything.

Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:29]: I think "meta packages" (Microsoft.Net.Compilers.Toolset are the simplest example where nothing is being used from the direct reference, but transitive references are being used.

**Comment by JoeRobich** [comment]
@CyrusNajmabadi Not sure, but were you thinking that from the VisualStudio layer we have access to the PackageUninstallService as well as the ability to manipulate the VsProject project reference without relying on the Project System for this work?

File: src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs (hunk 1 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/IUnusedReferencesService.cs:18]: Removed

File: src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.UnusedReferences
+{
+    internal interface IReferenceCleanupService : IWorkspaceService
+    {
+        /// <summary>
+        /// Gets the current selected TargetFramework for the specified project.
+        /// </summary>
+        string GetTargetFramworkMoniker(ProjectId projectId);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:19]: Framwork

File: src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// For the given project, returns the full path to the project.assets.json file
+        /// typically generated in the intermediate output path by a NuGet restore.
+        /// </summary>
+        Task<string> GetProjectAssetsFilePathAsync(
+            string projectPath,
+            string targetFramework,
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:27]: should this be `targetFrameworkMoniker`?  is there a difference between the two?  i'm trying to reconcile this name with the name of the method above.

File: src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:19]: is there a reason we need to know the current selecte tfm, as opposed to all the tfms the project supports?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:19]: should this be async?

File: src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        Task<ImmutableArray<Reference>> GetProjectReferencesAsync(
+            string projectPath,
+            string targetFramework,
+            CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Updates the project’s references by adding, removing or marking references as
+        /// TreatAsUsed in the project file.
+        /// </summary>
+        Task<bool> UpdateReferencesAsync(
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:43]: what is the 'bool' return value?

File: src/Features/Core/Portable/UnusedReferences/Reference.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.UnusedReferences
+{
+    internal class Reference
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/Reference.cs:7]: so we already have MetadataReference.  I'm wary of reserving the name Reference for this type.  i'm strugglign with a better name though.  Maybe ReferenceDescription or ReferenceInfo.

File: src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj [hunk 9]
```diff
@@ -24,15 +24,12 @@
     </PackageDescription>
   </PropertyGroup>
   <ItemGroup Label="NuGet">
-    <NuspecProperty Include="MicrosoftCSharpVersion=$(MicrosoftCSharpVersion)"/>
-    <NuspecProperty Include="MicrosoftVisualStudioCompositionVersion=$(MicrosoftVisualStudioCompositionVersion)"/>
-    <NuspecProperty Include="SystemThreadingTasksDataflowVersion=$(SystemThreadingTasksDataflowVersion)"/>
+    <NuspecProperty Include="MicrosoftCSharpVersion=$(MicrosoftCSharpVersion)" />
```
Comment (JoeRobich) [inline src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj:27]: I need to revert this file's changes.

File: src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    internal class UnusedReferencesService : IUnusedReferencesService
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public UnusedReferencesService()
+        {
+        }
+
+        public async Task<ImmutableArray<Reference>> GetUnusedReferencesAsync(Project project, CancellationToken cancellationToken)
+        {
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs:26]: if this is real code, check first if project.SupportsCompilation

File: src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+                return ImmutableArray<Reference>.Empty;
+            }
+
+            var usedReferences = compilation.GetUsedAssemblyReferences(cancellationToken);
+            var unusedReferences = compilation.References.Except(usedReferences).ToImmutableArray();
+
+            var targetFrameworkMoniker = referenceUpdateService.GetTargetFramworkMoniker(project.Id);
+            var projectAssetsFilePath = await referenceUpdateService.GetProjectAssetsFilePathAsync(project.FilePath, targetFrameworkMoniker, cancellationToken).ConfigureAwait(false);
+
+            var projectAssetsText = File.ReadAllText(projectAssetsFilePath);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs:46]: 1. definitely has to be a protected region.  we have IOUtilities for thsi purpose.  
2. should also be done prior doing the expensive compilation work. 
3. What is the failure mode for GetUnusedReferencesAsync?  if we can't read IO, do we propagate that higher?  or do we catch and expose the result in way that gives this information to the caller.

File: src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+            return ImmutableArray<Reference>.Empty;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs:48]: i assume this is undon.e

File: src/Features/Core/Portable/UnusedReferences/UpdateAction.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UnusedReferences/UpdateAction.cs:17]: Couldn't we have TreatAsUsed and TreatAsUnused be two different update actions?

File: src/VisualStudio/Core/Def/Implementation/UnusedReferences/VisualStudioReferenceCleanupService.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+        public string GetTargetFramworkMoniker(ProjectId projectId)
+        {
+            if (_workspace.TryGetHierarchy(projectId, out var hierarchy) &&
+                hierarchy.TryGetTargetFrameworkMoniker((uint)VSConstants.VSITEMID.Root, out var targetFrameworkMoniker))
+            {
+                return targetFrameworkMoniker ?? string.Empty;
+            }
+
+            return string.Empty;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/UnusedReferences/VisualStudioReferenceCleanupService.cs:48]: need to doc failure mode for this API.

File: src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs (hunk 12 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs:48]: It is. If it is preferable to just commit the interfaces initially, I can remove.

File: src/Features/Core/Portable/UnusedReferences/Reference.cs (hunk 8 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/Reference.cs:7]: I like ReferenceInfo.

File: src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs (hunk 10 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs:26]: removing this class for now, but good to know for the future.

File: src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs (hunk 7 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:43]: Added a comment that it indicates success and changing the method to work against a single ReferenceUpdate at a time.

File: src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs (hunk 6 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:27]: Yup, renaming to match

File: src/Features/Core/Portable/UnusedReferences/UpdateAction.cs (hunk 13 above)
Comment (JoeRobich) [reply src/Features/Core/Portable/UnusedReferences/UpdateAction.cs:17]: Sure, splitting into two

File: src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+{
+    internal interface IReferenceCleanupService : IWorkspaceService
+    {
+        /// <summary>
+        /// Gets the current selected TargetFrameworkMoniker for the specified project.
+        /// </summary>
+        string GetTargetFrameworkMoniker(ProjectId projectId);
+
+        /// <summary>
+        /// For the given project, returns the full path to the project.assets.json file
```
Comment (JoeRobich) [inline src/Features/Core/Portable/UnusedReferences/IReferenceCleanupService.cs:22]: A single project.assets.json is generated containing the dependency information for all TFMs.

**Comment by JoeRobich** [comment]
@ocallesp & @CyrusNajmabadi I've updated this PR based on feedback and it is ready for another review.

**Review by CyrusNajmabadi — APPROVED** [review]
i think this is good enough to move forward with integration.  we may run into issues later, but we can cross that bridge then :)

**Comment by JoeRobich** [comment]
It appears integration test running is broken in this feature branch because it hasn't been updated and the Integration machines are now running VS 16.8 P2 which had breaking changes.

**Pull request merged**